// @jupyterlab/testutils/lib/index.js
"use strict";
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Mock = exports.flakyIt = exports.dismissDialog = exports.dangerDialog = exports.acceptDialog = exports.waitForDialog = exports.initNotebookContext = exports.createFileContextWithKernel = exports.createFileContext = exports.createSession = exports.createSessionContext = exports.sleep = exports.framePromise = exports.isFulfilled = exports.signalToPromise = exports.signalToPromises = exports.expectFailure = exports.testEmission = exports.JupyterServer = exports.defaultRenderMime = exports.NBTestUtils = void 0;
var notebook_utils_1 = require("./notebook-utils");
Object.defineProperty(exports, "NBTestUtils", { enumerable: true, get: function () { return notebook_utils_1.NBTestUtils; } });
var rendermime_1 = require("./rendermime");
Object.defineProperty(exports, "defaultRenderMime", { enumerable: true, get: function () { return rendermime_1.defaultRenderMime; } });
var start_jupyter_server_1 = require("./start_jupyter_server");
Object.defineProperty(exports, "JupyterServer", { enumerable: true, get: function () { return start_jupyter_server_1.JupyterServer; } });
var common_1 = require("./common");
Object.defineProperty(exports, "testEmission", { enumerable: true, get: function () { return common_1.testEmission; } });
Object.defineProperty(exports, "expectFailure", { enumerable: true, get: function () { return common_1.expectFailure; } });
Object.defineProperty(exports, "signalToPromises", { enumerable: true, get: function () { return common_1.signalToPromises; } });
Object.defineProperty(exports, "signalToPromise", { enumerable: true, get: function () { return common_1.signalToPromise; } });
Object.defineProperty(exports, "isFulfilled", { enumerable: true, get: function () { return common_1.isFulfilled; } });
Object.defineProperty(exports, "framePromise", { enumerable: true, get: function () { return common_1.framePromise; } });
Object.defineProperty(exports, "sleep", { enumerable: true, get: function () { return common_1.sleep; } });
Object.defineProperty(exports, "createSessionContext", { enumerable: true, get: function () { return common_1.createSessionContext; } });
Object.defineProperty(exports, "createSession", { enumerable: true, get: function () { return common_1.createSession; } });
Object.defineProperty(exports, "createFileContext", { enumerable: true, get: function () { return common_1.createFileContext; } });
Object.defineProperty(exports, "createFileContextWithKernel", { enumerable: true, get: function () { return common_1.createFileContextWithKernel; } });
Object.defineProperty(exports, "initNotebookContext", { enumerable: true, get: function () { return common_1.initNotebookContext; } });
Object.defineProperty(exports, "waitForDialog", { enumerable: true, get: function () { return common_1.waitForDialog; } });
Object.defineProperty(exports, "acceptDialog", { enumerable: true, get: function () { return common_1.acceptDialog; } });
Object.defineProperty(exports, "dangerDialog", { enumerable: true, get: function () { return common_1.dangerDialog; } });
Object.defineProperty(exports, "dismissDialog", { enumerable: true, get: function () { return common_1.dismissDialog; } });
var flakyIt_1 = require("./flakyIt");
Object.defineProperty(exports, "flakyIt", { enumerable: true, get: function () { return flakyIt_1.flakyIt; } });
const Mock = __importStar(require("./mock"));
exports.Mock = Mock;
